R/vol-surface-class.R

```diff
@@ -0,0 +1,173 @@
+    !is.unsorted(as.Date(unique(x$vol_quotes$maturity), format = "%Y%m%d"), na.rm = F),
+    length(x$vol_quotes$strike) == length(unique(x$vol_quotes$strike)) *
+      length(unique(x$vol_quotes$maturity)),
+    is.character(x$ticker),
+    is.Interpolation(x$interpolation)
+  )
+  x
+}
+
+
+#' Inherits from VolSurface
+#'
+#' Checks whether object inherits from `VolSurface` class
+#'
+#' @param x an R object
+#' @return `TRUE` if `x` inherits from the `VolSurface` class; otherwise `FALSE`
+#' @export
+is.VolSurface <- function(x) {
+  inherits(x, "VolSurface")
+}
+
+
+#' @export
+format.VolSurface <- function(x, ...){
+  cat(paste0("<VolSurface> @", format(x$reference_date, "%e %B %Y")), "\n",
+   paste0(x$ticker, "  ", format(x$interpolation)))
+}
+
+#' @export
+print.VolSurface  <- function(x, ...){
+
+  cat(format(x), "\n")
+  print(x$vol_quotes)
+
+}
+
+
+# VolSurface methods --------------------------------
+
+
+#' Interpolate a `VolSurface` object.
+#'
+#' This method is used to interpolate a `VolSurface` object at multiple points of
+#' the plane. The interpolation depends on the type of the surface, if the vols are
+#' given by strikes, delta, moneyness.
+#'
+#' @param x object of class `VolSurface` to be interpolated.
+#' @param at indicates the coordinates at which the interpolation is performed.
+#' `at` should be given as a named list of length two, where the two members are
+#' vectors with the same length and are named `term` and `strike`. e.g.
+#' list(term = c(1, 2), strike = c(72, 92)).
+#' @param ... unused in this model.
+#' @return `numeric` vector with length equal to the length of `at` members.
+#' @examples vol_surface <- fmbasics::build_vol_surface()
+#' interpolation_points <- list(term = c(as.Date("2020-03-31"), as.Date("2021-03-31")),
+#' strike = c(40, 80))
+#' implied_vols <- fmbasics::interpolate(x = vol_surface, at  = interpolation_points)
+#' @family interpolate functions
+#' @export
+interpolate.VolSurface <- function(x, at, ...) {
+  assertthat::assert_that(
+    all.equal(names(at), c("term", "strike")),
+    length(at$term) == length(at$strike),
+    assertthat::is.date(at$term),
+    is.numeric(at$strike)
+  )
+
+  if (is.LinearTimeVarInterpolation(x$interpolation) & colnames(x$vol_quotes)[2] == "strike") {
+    vols <- x$vol_quotes$value
+    variance <- vols^2
+    tenors <- as.Date(x$vol_quotes$maturity, format = "%Y%m%d")
+    tenors_yf <- fmdates::year_frac(
+      date1 = x$reference_date, date2 = tenors,
+      day_basis = "act/365"
+    )
+
+    time_variance <- variance * tenors_yf
+
+    imp_vol <- rep(NA, length(at$term))
+
+
+    for (i in seq_along(at$term)) {
+      tt <- at$term[i]
+      K <- at$strike[i]
+      smile <- rep(NA, length(unique(x$vol_quotes$strike)))
+      tt_yf <- fmdates::year_frac(
+        date1 = x$reference_date, date2 = tt,
+        day_basis = "act/365"
+      )
+      for (k in 1:length(smile)) {
+        g <- stats::approxfun(
+          x = unique(tenors_yf),
```

Comment:
This seems to be the way you are trying to deal with `vol_quotes` being a tidy tibble. See next comment - is this really right?

Reply:
The mains conditions for the vols data are sorted dimension and for the data to be a rectangular data.